File: solutions/opus-1/opus/perl/lcs.pl
use strict;
use warnings;

my $a = <STDIN>;
my $b = <STDIN>;
$a = '' unless defined $a;
$b = '' unless defined $b;
$a =~ s/[\r\n]+$//;
$b =~ s/[\r\n]+$//;

my @x = split //, $a;
my @y = split //, $b;
my $m = scalar @y;

my @prev = (0) x ($m + 1);
for my $c (@x) {
    my @cur = (0);
    my $left = 0;
    for my $j (1 .. $m) {
        my $v;
        if ($y[$j - 1] eq $c) {
            $v = $prev[$j - 1] + 1;
        } else {
            my $up = $prev[$j];
            $v = $up > $left ? $up : $left;
        }
        push @cur, $v;
        $left = $v;
    }
    @prev = @cur;
}
print $prev[$m], "\n";

File: solutions/opus-1/opus/rust/lcs.rs
use std::io::{self, Read};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut lines = input.split('\n');
    let a: Vec<u8> = lines.next().unwrap_or("").trim().bytes().collect();
    let b: Vec<u8> = lines.next().unwrap_or("").trim().bytes().collect();
    let m = b.len();
    let mut prev = vec![0u32; m + 1];
    let mut cur = vec![0u32; m + 1];
    for &ca in &a {
        for j in 1..=m {
            cur[j] = if ca == b[j - 1] {
                prev[j - 1] + 1
            } else {
                prev[j].max(cur[j - 1])
            };
        }
        std::mem::swap(&mut prev, &mut cur);
    }
    println!("{}", prev[m]);
}
